src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -88,7 +88,12 @@ public static async ValueTask<SslStream> EstablishSslConnectionAsync(SslClientAu
                     throw CancellationHelper.CreateOperationCanceledException(e, cancellationToken);
                 }
 
-                throw new HttpRequestException(SR.net_http_ssl_connection_failed, e);
+                HttpRequestException ex = new HttpRequestException(SR.net_http_ssl_connection_failed, e);
+                if (request.IsExtendedConnectRequest)
+                {
+                    ex.Data["HTTP2_ENABLED"] = false;
```

Comment:
It seems a bit strange given that the code seems to be about SSL in general, not HTTP/2 specific; and it is also possible to have plaintext HTTP/2.

Reply:
We might want to have another name for Data here; I would also expect if SSL is not supported for the server, but we want a secure websocket connection ("wss") -- it doesn't make sense to try 1.1 after that because it would fail again? Unless I miss something...

Reply:
`sslOptions` is HTTP/2 specific in this method, we try to get it for exact HTTP/2 without knowledge that downgrade is possible -- it is out of scope for HttpClient

Reply:
Why it doesn't make sense to try 1.1? Secure websocket connection is supported for it

Reply:
I did some code investigation and now I understand, the SSL connection was trying to be established for strictly ALPN="h2", so ALPN="http/1.1" was not allowed by client and that's why SSL connection failed

Reply:
Maybe worth adding a comment about it